teps/0080-support-domainscoped-parameterresult-names.md

```diff
@@ -0,0 +1,165 @@
+creation-date: '2021-08-19'
+last-updated: '2021-08-19'
+authors:
+- '@mattmoor'
+---
+
+# TEP-0080: Support domain-scoped parameter/result names
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes/Caveats (optional)](#notescaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [User Experience (optional)](#user-experience-optional)
+  - [Performance (optional)](#performance-optional)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [Implementation Pull request(s)](#implementation-pull-request-s)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+Allow Task and Pipeline authors to expose parameters and results that allow
+`.` characters.
+
+> _This TEP was originally raised [here](https://github.com/tektoncd/pipeline/issues/3590)
+as an issue for discussion._
+
+## Motivation
+
+The motivation of this work is to *enabl* us to establish conventions around the
+definition of parameters and results that may have a deeper meaning for
+higher-level systems without a high-degree of accidental naming collisions.
+
+The pursuit of `.` is specifically to follow the precedents set by many other systems
+(including Kubernetes labels, CloudEvent types, Java import paths, ContainerD plugins,
+and many more) of using a Domain-scoped naming convention to "claim" a segment of
+names for the owner's exclusive use.  This is of course conventional, but in practice
+is generally good enough.
+
+With this work, higher-level systems could start to define Task/Pipeline "interfaces"
+or leverage partial signature matching to enable special semantics around certain
+signatures, without (significant) concerns around matching something accidentally.
+
+> _It is notable that with @bobcatfish [TEP for stronger
+typing](https://github.com/tektoncd/community/pull/479) that these conventions should
+also establish the types associated with those parameter names, and might form a good
+case for SchemaRefs._
+
+### Goals
+
+Enable conventions to be established around the use of domain-scoped names as a way
+for the domain owner to "own" the definition of what those parameters and results are
+for and how they will be used.
+
+### Non-Goals
+
+It is not a goal of this TEP to establish these conventions, or begin to define any
+"well known" parameters or results owned by Tekton (e.g. `dev.tekton.foo.bar`).
+
+
+### Use Cases (optional)
+
+See the [original issue](https://github.com/tektoncd/pipeline/issues/3590) for some
+of these.
```

Comment:
is it possible to go into more detail around these? i dont totally understand the use cases myself - and I'd like to understand the use cases we have for this that wouldn't be satisfied by having dictionary/object support ([TEP-0075](https://github.com/tektoncd/community/pull/479)) ?

Reply:
They are completely unrelated.  The motivation is to be able to define parameters to Tasks and Pipelines that enable a level of unambiguous matching of certain semantic characteristics.  For instance, if a Task or Pipeline were parameterized by some workspace setup logic, being able to unambiguously identify that semantic through a highly scoped parameter name would be extremely useful.  Ultimately this isn't about teaching tektoncd/pipelines any new tricks, but providing enough leeway that higher-level tooling (incl. potentially `tkn`) can create higher-level experiences.

I'm currently doing this extensively in `mink`, but with `-`, which leaves a lot to be desired.  I enumerated a large number of potential use cases here: https://github.com/tektoncd/community/pull/504

Reply:
I'm also happy to jump on a quick call if it'd be easier to talk through synchronously.

Reply:
>  For instance, if a Task or Pipeline were parameterized by some workspace setup logic, being able to unambiguously identify that semantic through a highly scoped parameter name would be extremely useful.

I'm still not understanding why having dictionary parameter support wouldn't give you that.

In #504 you show this example:

```
      params:
        - name: dev.mink.kontext
          description: A self-extracting container image of source
```

If we had dictionary support (eventually nested dictionary support) that could be expressed as:

```
      params:
       - name: dev
         description: A self-extracting container image of source
         type: object
         properties:
         mink: {
              type: object
              properties: {
                 kontext: {
                     type:string
                }
              } 
          }

```

It's certainly a lot longer, so maybe it's the verbosity that makes you want to prefer encoding the namespacing into the string? The above structure would let you define other properties as part of your mink object as well - and as @wlynch had mentioned we could have predefined schemas for known types (like "mink params").

Reply:
> the structure of the dictionary does express the same information

To a machine capable of deep structural analysis of the type? yes for this example.  At a glance, absolutely not (thus the term "obfuscate").

I also don't think keys like `openshift` and `mink` are unambiguous enough that we should expect common users to avoid defining them, where it's a pretty well-worn precedent that domains are the purview of the domain owner.


I also had a thought this morning, which I think will help shine a light on the confusion and distinction I am trying to draw between namespaced field names and decomposing those same fields into objects themselves ([see](https://github.com/tektoncd/community/pull/479/files#r697888394)).  If the plan with TEP 75 is to allow for structured params and results following JSON structure, then this TEP devolves to handling fields within those same objects with `.` in them (and JSON fields can have a lot more characters Tekton likely doesn't support today!).  My point is that the same trick doesn't work here (from my linked comment):

```yaml
         name: line
         type: object
         properties:
           "knative.dev/controller": { type: string}
           "knative.dev/key": { type: string}
           ...
```

... this TEP is really just a degenerate case of this in a world where only primitive types are supported.

Reply:
hmm so what I'm taking away from this is that a different way of looking at this proposal might be something like: as we expand support for json based params, we should look at supporting more of the characters that json considers valid which we currently don't, including `.` (and `/` in the examples you added)

seems reasonable to me!

i tried to see if there was any reason why we are limiting the characters the way we are and it seems to come down to https://github.com/tektoncd/pipeline/pull/472 where @vdemeester says:

> I think this is a good "base" for variable name but might be too restrictive though

so it seems like it was arbitrary and expanding to include `.` makes sense (based on your example above would you want to include `/` as well? "knative.dev/controller" is a bit more readable for me personally than "dev.knative.controller" but that's just me)